pages/pipelines/defining_steps.md.erb

```diff
@@ -270,6 +270,24 @@ steps:
 ```
 {: codeblock-file="pipeline.yml"}
 
+## Dynamic pipeline templates
+
+If you need the ability to use pipelines from a central catalog, or enforce certain configuration rules, you can either use dynamic pipelines and the [`pipeline upload`](/docs/agent/v3/cli-pipeline#uploading-pipelines) command to make this happen, or [write custom plugins](/docs/plugins/writing) and share them across your organization.
+
+To use dynamic pipelines and the pipeline upload command, you'd make a pipeline that looks something like this:
+
+```yml
+steps:
+  - command: enforce-rules.sh | buildkite-agent pipeline upload
+    label: "\:pipeline\: Upload"
+```
+
+Where each team defines their steps in `team-steps.yml`, and your templating logic is in `enforce-rules.sh` (which can be written in any language that can pass YAML to the pipeline upload).
+
+In `enforce-rules.sh` you can add steps to the YAML, require certain versions of dependencies or plugins, or implement any other logic you can program. Depending on your use case you might want to get `enforce-rules.sh` from an external catalog instead of committing it to the team repository.
+
+[Hasura.io](https://hasura.io) blogged about using [dynamic templates and pipelines](https://hasura.io/blog/what-we-learnt-by-migrating-from-circleci-to-buildkite/#dynamic-pipelines) to replace their YAML configuration with Go and some shell scripts.
```

Comment:
```suggestion
For some inspiration, you can read a blog post by [Hasura.io](https://hasura.io) about using [dynamic templates and pipelines](https://hasura.io/blog/what-we-learnt-by-migrating-from-circleci-to-buildkite/#dynamic-pipelines) to replace their YAML configuration with Go and some shell scripts.
```

Reply:
"For some inspiration" can be replaced with something better, but I'd really avoid having an external link in the very first word of the sentence - as a reader, I'd like to have some context as to what I'm suggested to click.

Reply:
```suggestion
See how [Hasura.io](https://hasura.io) used [dynamic templates and pipelines](https://hasura.io/blog/what-we-learnt-by-migrating-from-circleci-to-buildkite/#dynamic-pipelines) to replace their YAML configuration with Go and some shell scripts.
```